src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java

```diff
@@ -0,0 +1,976 @@
+            float x = (float) con;
+            short sx = (short)x;
+            if (x == sx) {
+                if (sx >= 0 && sx <= 2) {
+                    mv.visitInsn(Opcodes.FCONST_0 + (int) sx);
+                } else {
+                    emitIconstInsn((int) x);
+                    mv.visitInsn(Opcodes.I2F);
+                }
+                return;
+            }
+        }
+        if (con instanceof Double) {
+            double x = (double) con;
+            short sx = (short)x;
+            if (x == sx) {
+                if (sx >= 0 && sx <= 1) {
+                    mv.visitInsn(Opcodes.DCONST_0 + (int) sx);
+                } else {
+                    emitIconstInsn((int) x);
+                    mv.visitInsn(Opcodes.I2D);
+                }
+                return;
+            }
+        }
+        if (con instanceof Boolean) {
+            emitIconstInsn((boolean) con ? 1 : 0);
+            return;
+        }
+        // fall through:
+        mv.visitLdcInsn(con);
+    }
+
+    private void emitIconstInsn(int cst) {
+        if (cst >= -1 && cst <= 5) {
+            mv.visitInsn(Opcodes.ICONST_0 + cst);
+        } else if (cst >= Byte.MIN_VALUE && cst <= Byte.MAX_VALUE) {
+            mv.visitIntInsn(Opcodes.BIPUSH, cst);
+        } else if (cst >= Short.MIN_VALUE && cst <= Short.MAX_VALUE) {
+            mv.visitIntInsn(Opcodes.SIPUSH, cst);
+        } else {
+            mv.visitLdcInsn(cst);
+        }
+    }
+
+    private void emitLoad(Class<?> type, int index) {
+        emitLoad(BasicType.of(type), index);
+    }
+
+    private void emitLoad(BasicType type, int index) {
+        int opcode = loadOpcode(type);
+        mv.visitVarInsn(opcode, index);
+    }
+
+    private static int loadOpcode(BasicType type) throws InternalError {
+        return switch (type) {
+            case Z, B, S, C, I -> ILOAD;
+            case J -> LLOAD;
+            case F -> FLOAD;
+            case D -> DLOAD;
+            case L -> ALOAD;
+        };
+    }
+
+    private void emitStore(BasicType type, int index) {
+        int opcode = storeOpcode(type);
+        mv.visitVarInsn(opcode, index);
+    }
+
+    private static int storeOpcode(BasicType type) throws InternalError {
+        return switch (type) {
+            case Z, B, S, C, I -> ISTORE;
+            case J -> LSTORE;
+            case F -> FSTORE;
+            case D -> DSTORE;
+            case L -> ASTORE;
+        };
+    }
+
+    private void emitReturn(Class<?> type) {
+        int opcode = returnOpcode(BasicType.of(type));
+        mv.visitInsn(opcode);
+    }
+
+    private static int returnOpcode(BasicType type) {
+       return switch (type) {
+            case Z, B, S, C, I -> IRETURN;
+            case J -> LRETURN;
+            case F -> FRETURN;
+            case D -> DRETURN;
+            case L -> ARETURN;
+        };
+    }
+
+    private enum BasicType {
+        Z, B, S, C, I, J, F, D, L;
+
+        static BasicType of(Class<?> cls) {
```

Comment:
This seems a duplication of ASM Type class for no good reason, Type.getOpcode(ILOAD) or Type.getOpcode(IRETURN) gives you the proper variant opcode

Reply:
Didn't know about that. Neat!

Reply:
Removed the `BasicType` enum, switched to using `Type` and `Type.getSort()`.